### ISSUE TITLE
New cop for `#dig` followed by `#first`

### ISSUE THREAD

**Issue by lovro-bikic** [rubocop/rubocop#13487]
## Is your feature request related to a problem? Please describe.

I've often seen code like this:
```ruby
x.dig(:foo, :bar).first
x.dig(:foo, :bar)&.first
```
which could usually be replaced with:
```ruby
x.dig(:foo, :bar, 0)
```

This is of course a stylistic choice; an explicit `#first` might look better to someone. However, I've also seen `#first` then followed by another `#dig`:
```ruby
x.dig(:foo, :bar).first.dig(:baz)
x.dig(:foo, :bar)&.first&.dig(:baz)
```
which would perhaps work better as a single dig:
```ruby
x.dig(:foo, :bar, 0, :baz)
```

(see Additional context section for info on how this would be achieved)

## Describe the solution you'd like

A new cop to detect usage of `dig(*a)&.first / dig(*a).first` and autocorrect to `dig(*a, 0)`.

## Additional context

The case `dig.first.dig` would work in conjuction with [this cop for chained digs](https://github.com/rubocop/rubocop/issues/13486), e.g.:
```ruby
x.dig(:foo, :bar)&.first&.dig(:baz)
# autocorrected to by this cop:
x.dig(:foo, :bar, 0)&.dig(:baz)
# autocorrected to by cop for chained digs:
x.dig(:foo, :bar, 0, :baz)
```

Here are [search results](https://github.com/search?q=%2F%5Cbdig%5Cb%5C%28.%2B%3F%5C%29%26%3F%5C.%5Cbfirst%5Cb%2F+language%3ARuby&type=code) which show the impact this cop would have, and [these search results](https://github.com/search?q=%2F%5Cbdig%5Cb%5C%28.%2B%3F%5C%29%26%3F%5C.%5Cbfirst%5Cb.*%26%3F%5C.dig%5Cb%2F+language%3ARuby&type=code) show the impact of this cop in conjuction with the other suggested cop.

This cop would probably be unsafe, `.first` might not yield the same result as `.dig(0)` if the receiver implements custom `#first`.

**Comment by dvandersluis** [comment]
What about `dig(:foo, :bar).last` => `dig(:foo, :bar, -1)`?

**Comment by lovro-bikic** [comment]
👍 a good point, it didn't occur to me (I haven't used it before tbh). I guess that should be supported as well.

I'm also not sure if there should be support for [additional methods added by ActiveSupport](https://github.com/rails/rails/blob/d3b2fcb4ffee825f8e9ad8f4bacc859978aebaed/activesupport/lib/active_support/core_ext/array/access.rb#L55-L78) when `ActiveSupportExtensionsEnabled` option is enabled?

**Comment by koic** [comment]
Um, I have doubts about whether `x.dig(:foo, :bar, 0)` is more readable than `x.dig(:foo, :bar).first`.
From a performance perspective, `first` is also faster, which raises doubts about this proposal.

```console
$ cat dig.rb
#!/usr/local/bin/ruby

require 'benchmark/ips'

Benchmark.ips do |x|
  h = {foo: {bar: [42]}}

  x.report('first') { h.dig(:foo, :bar).first }
  x.report('0')     { h.dig(:foo, :bar, 0) }

  x.compare!
end

$ ruby dig.rb
ruby 3.3.6 (2024-11-05 revision 75015d4c1f) [x86_64-darwin23]
Warming up --------------------------------------
               first     1.059M i/100ms
                   0   959.494k i/100ms
Calculating -------------------------------------
               first     10.359M (± 2.4%) i/s   (96.54 ns/i) -     51.877M in   5.011460s
                   0      9.157M (± 1.0%) i/s  (109.20 ns/i) -     46.056M in   5.029982s

Comparison:
               first: 10358506.6 i/s
                   0:  9157197.7 i/s - 1.13x  slower
```

**Comment by lovro-bikic** [comment]
> I have doubts about whether x.dig(:foo, :bar, 0) is more readable than x.dig(:foo, :bar).first.

I agree, that's why I put a note in the description that this case doesn't necessarily make sense. The primary motivation for this cop was to style chained digs broken up by `first`, `last`, `second`, etc., into a single dig:
```ruby
response.dig(:results)&.first&.dig(:locations).first&.dig(:latLng)
# becomes
response.dig(:results, 0, :locations, 0, :latLng)
```

This example was taken from [here](https://github.com/search?q=%2F%5Cbdig%5Cb%5C%28.%2B%3F%5C%29%26%3F%5C.%5Cbfirst%5Cb.*%26%3F%5C.dig%5Cb%2F+language%3ARuby&type=code).

**Comment by dvandersluis** [comment]
If this were made a cop it wouldn't be one I'd want to use, I don't like using ints inside of `dig`.

That being said, it clearly is a style that has <a href="https://github.com/search?q=lang%3Aruby+%2Fdig%5C%28%5B%5E%29%5D%2B%28%2C%5Cs*0%7C0%2C%29%5B%5E%29%5D%2B%5C%29%2F&type=code">some use</a>.
